_posts/2016-03-12-make-your-data-obey.markdown

```diff
@@ -0,0 +1,187 @@
+
+**[Obey](https://github.com/TechnologyAdvice/obey) was developed specifically to address, and provide solutions for, data modelling and validation for modern web applications developed in JavaScript.**
+
+## Diving In
+
+Let's start with a simple example; validating a user account. There are some core data points we want to ensure, so a model designating these is a good starting point:
+
+{% highlight javascript %}
+import obey from 'obey'
+
+const user = obey.model({
+    id: { type: 'uuid', required: true },
+    email: { type: 'email', required: true },
+    password: { type: 'string', require: true }
+})
+{% endhighlight %}
+
+The above should all be fairly easy to understand. Now let's put it into action, saving to a datasource. Let's pretend this is going into a datasource using a CRUD lib, already setup, called `db`. Using the model above:
+
+{% highlight javascript %}
+import db from './db'
+
+user.validate({
+    id: 'fe6f80c6-c530-4f10-ad0c-89f0f68a879f',
+    email: 'jdoe@email.com',
+    password: 'p@55w0rd'
+})
+// Assumes a db.create method which accepts an object as argument
+.then(db.create).then(res => {
+    console.log('User Created!', res)
+})
+.catch(err => {
+    console.log('Failed!', err)
+})
+{% endhighlight %}
+
+Overall pretty simple stuff. We validate the model, then create the user (or catch the error). But what about that `id`? We probably want to have that generated automatically. Yes, most databases auto-generate unique id's, but play along; this is a perfect example.
+
+## Creators
+
+Obey supplies `creators` for generating data using methods that can be predefined. For the `id` we'll use the [node-uuid](https://www.npmjs.com/package/node-uuid) module. Let's revisit the model:
+
+{% highlight javascript %}
+import obey from 'obey'
+// Import the node-uuid lib
+import uuid from 'node-uuid'
+
+// Add a creator that returns a v4 UUID if one isn't supplied
+obey.creator('uuidCreator', () => uuid.v4())
+
+const user = obey.model({
+    // The id now specifies that the creator 'uuid' should be used if no value is present
+    id: { type: 'uuid', creator: 'uuidCreator' required: true },
+    email: { type: 'email', required: true },
+    password: { type: 'string', require: true }
+})
+{% endhighlight %}
+
+So, now we just pass in the `email` and `password` and our `id` is auto-generated.
+
+But that `password` is still not cool; we don't want to save that in plain-text...
+
+## Modifiers
+
+Obey wasn't just built for validation, but also to support coercion and modification. Modifiers allow for doing this in a clean, simple method. Let's revisit the model again:
+
+{% highlight javascript %}
+import obey from 'obey'
+import uuid from 'node-uuid'
+// Import the crypto module
+import crypto from 'crypto'
+
+obey.creator('uuid', () => uuid.v4())
+
+// Add a modifier that returns md5 encrypted value
+obey.modifier('encrypt', val => crypto.createHash('md5').update(val).digest('hex'))
```

Comment:
You'll earn huge cred here and avoid people attacking irrelevant semantics of you include argon2. The code can be just as brief-- see userhub/src/hashStrategies/argon2.js